Default to MPAv2 with changes and adjust tests we accept as graceful degradation

kmcgrady opened the pull request:
## Describe your changes

This change switches the Default strategy on both the PagesManager and AppNavigation. We discussed cases where this may produce breaking behavior, but we would call these graceful degradation.

This effectively makes the V1 strategy dead code. A follow up PR will be done to remove the dead code and refactor so it's a little bit cleaner.

## Testing Plan

- Tests should pass

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

lukasmasuch committed:
Default to MPAv2 with changes and adjust tests we accept will break

kmcgrady committed:
Update snapshots

kmcgrady commented on e2e_playwright/multipage_apps/mpa_configure_sidebar_test.py:
```diff
@@ -100,7 +100,7 @@ def test_page_link_href(
     page_links = app.get_by_test_id("stPageLink-NavLink")
 
-    expect(page_links.nth(0)).to_have_attribute("href", "mpa_configure_sidebar")
+    expect(page_links.nth(0)).to_have_attribute("href", "")
```
This is the default page which we supply an empty pathname

kmcgrady committed:
Update sort page key

raethlein commented on lib/streamlit/runtime/scriptrunner/script_runner.py:
```diff
@@ -116,6 +117,43 @@ class ScriptRunnerEvent(Enum):
 """
 
 
+def _mpa_v1(main_script_path: str):
```
nit: I think a comment would be beneficial here explaining that the function transforms the pages folder into the MPAv2 representation

kmcgrady replied:
Updated

raethlein commented on lib/streamlit/elements/widgets/button.py:
```diff
@@ -873,11 +873,12 @@ def _page_link(
                 if requested_page == full_path:
                     if label is None:
-                        page_link_proto.label = page_name.replace("_", " ")
+                        page_link_proto.label = page_name
```
Why don't we have to replace underscores anymore now? Just want to make sure we capture a little bit of context here in the comments in case we have to reason about it again in the future

kmcgrady replied:
Since every page is being instantiated from `st.Page` now, the pages in the page manager are set correctly. [This line](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/navigation/page.py#L213) does the same thing.

Likely there was a bug where we use v2, and the page title had an underscore in it it was being stripped. But it was not found because (1) page titles with _ is not common and (2) the workaround was to leverage `st.Page`

raethlein commented on lib/streamlit/runtime/scriptrunner/script_runner.py:
```diff
@@ -588,6 +626,7 @@ def code_to_exec(code=code, module=module, ctx=ctx, rerun_data=rerun_data):
                                 pass
 
                     else:
+                        _mpa_v1(self._main_script_path)
```
This check seems to happen on every rerun request. It looks like right now the result of `mpa_v1` is not cached, which could make the app pretty slow. Instead of executing it here, it might be better to execute it once during the runtime startup or similar and / or add such a caching check. In case we want to retain any hot-reload behavior, we would need to find a way to invalidate the cache though 🤔

lukasmasuch replied:
I think that is comparable with a normal MPAv2 app. The main script + st.navigation is executed on every rerun. And I don't think there is any aspect in the `_mpa_v1` that has a big risk of slowing down an app. Maybe we could cache the request to the file system (to get the files in pages), but I think not caching it could be more of a feature than a bug since it would auto adjust to what pages are available on reruns. Probably something to do a performance measurement on how long it takes in a normal setup. But I suspect that its pretty negligible.

kmcgrady replied:
@lukasmasuch 's comment makes the most sense here. If we wanted to _really_ make this performant, we can add back the pages watcher and updated the files there.

I'll look into caching the `.exists()` call. This will make sure MPAv2 will be performant. And if there really are performance issues, we could redirect users to MPAv2

raethlein replied:
Then the bet is one 😄: I would suspect that the file system operation here adds measurable overhead with the lookup and filtering of files and then the sorting of them, which gets worse the more files are in the `/pages` folder.

lukasmasuch replied:
measurable, probably yes... significant not sure. The relevant distribution to perform well is probably up to 20 pages:

![image](https://github.com/user-attachments/assets/2c20053b-0880-4c1e-bca8-1967df9bcbb2)

Apps that have a lot more are very niche and should just switch to MPAv2.

lukasmasuch replied:
I did a quick check, executing this code with a folder with 20 py files takes ~1 second for 10k iterations (0.1 millisecond per iteration):

```python
import timeit
from pathlib import Path

import streamlit as st

# Select the folder that should be used for the pages:
PAGES_FOLDER = Path(__file__).parent / "shared"


def get_files():
    if PAGES_FOLDER.exists():
        # Read out the my_pages folder and create a page for every script:
        pages = PAGES_FOLDER.glob("*.py")
        pages = sorted([page for page in pages if page.name.endswith(".py")])
    return pages


st.write(timeit.timeit(get_files, number=10000))
```

sfc-gh-braethlein replied:
Awesome, thanks for backing it up with data! I surely overestimated the impact of file operations then!

raethlein replied:
Awesome, thanks for backing it up with data! I surely overestimated the impact of file operations then!

raethlein commented on e2e_playwright/__snapshots__/linux/mpa_v2_custom_theme_test/sidebar-nav-custom-theme[chromium].png:
sanity-check: is it expected that the icons are gone from the page titles?

kmcgrady replied:
This is a good catch. The answer is yes it is expected. This is due to the main script having an `st.navigation` call and `pages` folder. The test clicks on one of those pages. Since we do not have the navigation call, and we disregard the main script because we assume V1, it reverts back to the pages folder to render the sidebar.

So I think we consider this expected behavior.

That being said, the test is just weird as a result. I'll find out whether we can modify the test to better reflect what it's trying to test than have this nuance.

raethlein commented on frontend/app/src/util/AppNavigation.ts:
```diff
@@ -288,7 +288,13 @@ export class StrategyV2 {
-      ) ?? this.mainPage
+      ) ??
+      this.mainPage ?? {
+        // A navigation call should always be called before this, but we want to provide something
```
what do we provide here / what's the effect of returning an object with empty values?

kmcgrady replied:
I'll actually look into removing it. I added it originally because I wanted to resolve some JS tests, but the truth is we expect this to not be null and we should adjust the test to ensure the proper flow. To be more specific, I think this occurs in the `App.tsx` tests because we are now ensuring a `NAVIGATION` proto message is being sent which would update these tests.

kmcgrady replied:
Removing it is not an issue for now, so I will remove it for now.

raethlein commented on frontend/app/src/App.test.tsx:
```diff
@@ -1135,265 +1135,6 @@ describe("App", () => {
-        )
-      })
-
-      it("doesn't push a new history when the same page URL is already set", () => {
```
Do we not need these tests anymore because it cannot happen in MPAv2? Just trying to collect some more context in case someone has to look into this again in the future

kmcgrady replied:
Ah. I was originally going on a removing spree, but I stopped and decided to skip these in favor of "converting to MPA v2 tests" Will reintroduce and mark as skip and address this in the clean up PR.

lukasmasuch commented on lib/streamlit/navigation/page.py:
```diff
@@ -300,7 +300,7 @@ def run(self) -> None:
 
                 # We create a module named __page__ for this specific
                 # script. This is differentiate it from the `__main__` module
```
Is this comment still correct?

kmcgrady replied:
Will remove

kmcgrady committed:
Address PR comments

kmcgrady committed:
Address mypy error

kmcgrady committed:
cache pages folder existing

kmcgrady committed:
Fix has_pages_directory

kmcgrady committed:
Handle pages folder path resolution better

kmcgrady committed:
Handle v1->v2 more gracefully

kmcgrady committed:
Update snapshots

raethlein commented on lib/streamlit/runtime/scriptrunner/script_runner.py:
```diff
@@ -116,6 +117,50 @@ class ScriptRunnerEvent(Enum):
+
+    # Read out the my_pages folder and create a page for every script:
+    pages = PAGES_FOLDER.glob("*.py")
+    pages = sorted(
+        [page for page in pages if page.name.endswith(".py")], key=page_sort_key
+    )
+
+    # Use this script as the main page and
+    main_page = StreamlitPage(MAIN_SCRIPT_PATH, default=True)
+    all_pages = [main_page] + [
+        StreamlitPage(PAGES_FOLDER / page.name) for page in pages
+    ]
```
nit: maybe it make sense to loop only once through the `pages` here like:

```python
main_page = StreamlitPage(MAIN_SCRIPT_PATH, default=True)
all_pages = [main_page] + sorted([StreamlitPage(PAGES_FOLDER / page.name) for page in pages if page.name.endswith(".py")], key=page_sort_key])
```
where `page_sort_key` would need to be updated to take a `StreamlitPage` object.
But probably its overoptimization

raethlein reviewed (approved):
LGTM! Thanks for digging into this 🙌

lukasmasuch commented on lib/streamlit/runtime/pages_manager.py:
```diff
@@ -237,6 +232,13 @@ def __init__(
+        # for apps with a pages directory. We will keep this flag around
+        # for now to maintain the behavior for apps that were created with
+        # the pages directory feature.
+        PagesManager.uses_pages_directory = Path(
```
nit: maybe add as a comment the flag might be deactivated again if there is another `st.navigation` call indicating usage of MPAv2

lukasmasuch commented:
The code also looks good to me 👍  I did some testing (with the branch from https://github.com/streamlit/streamlit/pull/10458) and it mostly worked well, but there was some weirdness when there is a `pages` folder + MPAv2 and I tried to directly access an URL path: It usually resulted in a `Page not found` dialog but with the correct page being loaded in the background. Sometimes it completely switch back to MPAv1. 

What could help is global flag wether to use MPAv1 since the current flag seems to be session based. Which we put to true if there is a pages folder & to false on the first `st.navigation` call. But not sure if that is the best solution. 

We could also think about just logging a warning that its recommended not to use `pages` folder together with `st.navigation`, since in any case the combination is a slightly more degraded experience and could have some other unexpected issues.

kmcgrady committed:
Remove MPAv1 code (#10458)

## Describe your changes

This PR strips all of the MPAv1 code away including the strategy pattern
in both cases. It keeps AppNavigation in order to simplify code away
from `App.tsx`.

## Testing Plan

- Python and JS unit tests updated

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.
